<a href="https://colab.research.google.com/github/shubhamscifi/Instacart-Market-Basket-Analysis/blob/main/1.%20Problem%20and%20Data%20Overview.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **[Instacart Market Basket Analysis](https://www.kaggle.com/c/instacart-market-basket-analysis/)**

---
# **Problem Overview:**

Which products will an Instacart consumer purchase again?

1. Instacart is a grocery ordering and delivery app.
2.   Currently they use transactional data to develop models that predict which products a user will buy again, try for the first time, or add to their cart next during a session.
3.    The goal is to predict which previously purchased products will be in a user’s next order.
4. For each orderid in the test set, we should predict a space-delimited list of productids for that order. If we wish to predict an empty order, we should submit an explicit 'None' value. Note that some orders will have no reordered items. We may predict an explicit 'None' value for orders with no reordered items.
5. over 3 million grocery orders.
6. more than 200,000 Instacart users.
7. For each user, instacart provided between 4 and 100 of their orders, with the sequence of products purchased in each order.

---








## **Problem Definition:**
###    - Given order_id predict all the products that the user will reorder.

---



## **Business Objective:**
- Predicting reorders in advance is of huge business value for any delivery business, as this can significantly help in optimising the supply-chain management system.
- This can also be used to enhance the customer experience. eg:- Remind the customer to reorder a product.
- This can also be used to improve the recommender system.
- ML model can help here in predicting the reorders.
- Any delivery or retail business can get benefitted from the solution to this problem.

## **Business Constraints:**
- We want our model to perform equally good for all the orders.
- Predicting reorders of products that are reordered a lot is more important than the products that are reordered a few times.
- No strict latency requirements.
- Interpretability is partially important.
- Since recommending some products which a customer is not going to reorder will lead to the customer exploring new products, a few False Positives doesn't harm.
- False Negatives for highly reordered products can be costly.


---

## [**Dataset Description:**](https://gist.github.com/jeremystan/c3b39d947d9b88b3ccff3147dbcf6c6b#file-data_description-md)
<article class="markdown-body entry-content container-lg" itemprop="text"><p><code>orders</code> (3.4m rows, 206k users):</p>
<ul>
<li><code>order_id</code>: order identifier</li>
<li><code>user_id</code>: customer identifier</li>
<li><code>eval_set</code>: which evaluation set this order belongs in (see <code>SET</code> described below)</li>
<li><code>order_number</code>: the order sequence number for this user (1 = first, n = nth)</li>
<li><code>order_dow</code>: the day of the week the order was placed on</li>
<li><code>order_hour_of_day</code>: the hour of the day the order was placed on</li>
<li><code>days_since_prior</code>: days since the last order, capped at 30 (with NAs for <code>order_number</code> = 1)</li>
</ul>
<p><code>products</code> (50k rows):</p>
<ul>
<li><code>product_id</code>: product identifier</li>
<li><code>product_name</code>: name of the product</li>
<li><code>aisle_id</code>: foreign key</li>
<li><code>department_id</code>: foreign key</li>
</ul>
<p><code>aisles</code> (134 rows):</p>
<ul>
<li><code>aisle_id</code>: aisle identifier</li>
<li><code>aisle</code>: the name of the aisle</li>
</ul>
<p><code>deptartments</code> (21 rows):</p>
<ul>
<li><code>department_id</code>: department identifier</li>
<li><code>department</code>: the name of the department</li>
</ul>
<p><code>order_products__SET</code> (30m+ rows):</p>
<ul>
<li><code>order_id</code>: foreign key</li>
<li><code>product_id</code>: foreign key</li>
<li><code>add_to_cart_order</code>: order in which each product was added to cart</li>
<li><code>reordered</code>: 1 if this product has been ordered by this user in the past, 0 otherwise</li>
</ul>
<p>where <code>SET</code> is one of the four following evaluation sets (<code>eval_set</code> in <code>orders</code>):</p>
<ul>
<li><code>"prior"</code>: orders prior to that users most recent order (~3.2m orders)</li>
<li><code>"train"</code>: training data supplied to participants (~131k orders)</li>
<li><code>"test"</code>: test data reserved for machine learning competitions (~75k orders)</li>
</ul>
</article>


src: https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2

---



<div class="markdown-converter__text--rendered">
<h2>Submission File</h2>
<p>For each order_id in the test set, you should predict a space-delimited list of product_ids for that order. If you wish to predict an empty order, you should submit an explicit 'None' value. You may combine 'None' with product_ids. The spelling of 'None' is case sensitive in the scoring metric. The file should have a header and look like the following:</p>
<pre><code>order_id,products  
17,1 2  
34,None  
137,1 2 3  
etc.
</code></pre></div>

---

![instacart_Files.png](https://storage.googleapis.com/kaggle-forum-message-attachments/183176/6539/instacartFiles.png)
src: https://www.kaggle.com/c/instacart-market-basket-analysis/discussion/33128

---



## **Evaluation Metric**
* Submissions will be evaluated based on their 
<a href="https://scikit-learn.org/stable/modules/model_evaluation.html#:~:text=F%CE%B2(y%2Cy%5E)-,%22samples%22,-1%7CS%7C%E2%88%91s">Samples F1-score</a> i.e. average of F1-score for each order_id.

[](https://www.kaggle.com/c/instacart-market-basket-analysis/discussion/33357#:~:text=You%20take%20the%20average%20of%20all%20the%20F1%20scores%20across%20the%20order_ids)
    
    - Reason:
        - We want our model to give equal weightage to each order.
        - We want a model with good Precision and Recall both.

* Confusion Matrix will be used as secondary metric.
---



## System Specifications💻:

**Operating System:**

In [ ]:
!uname
!lsb_release -d

Linux
Description:	Ubuntu 18.04.5 LTS


**CPU Info:**

In [ ]:
!lscpu

Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  2
Core(s) per socket:  2
Socket(s):           1
NUMA node(s):        1
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7B12
Stepping:            0
CPU MHz:             2249.998
BogoMIPS:            4499.99
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-3
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignss

**RAM size:**

In [ ]:
!grep MemTotal /proc/meminfo    # 25.46 GB

MemTotal:       26698868 kB


**GPU info:**

In [ ]:
!nvidia-smi

Sat Jun 19 12:10:57 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 465.27       Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   44C    P8    10W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

## Downloading Dataset

In [ ]:
# Kaggle authentication
from getpass import getpass
import os

os.environ['KAGGLE_USERNAME'] = "shubhamscifi" #input('Enter kaggle username: ') # kaggle username
os.environ['KAGGLE_KEY'] = getpass('Enter Token: ') # kaggle api key

Enter Token: ··········


In [ ]:
# https://github.com/Kaggle/kaggle-api#download-competition-files
# fetching the instacart-market-basket-analysis dataset from kaggle.
!kaggle competitions download -c instacart-market-basket-analysis -p ./dataset

 72% 5.00M/6.90M [00:00<00:00, 19.5MB/s]
100% 6.90M/6.90M [00:00<00:00, 23.0MB/s]
  0% 0.00/220k [00:00<?, ?B/s]
100% 220k/220k [00:00<00:00, 69.9MB/s]
 93% 146M/157M [00:01<00:00, 98.1MB/s]
100% 157M/157M [00:01<00:00, 98.7MB/s]
  0% 0.00/804 [00:00<?, ?B/s]
100% 804/804 [00:00<00:00, 693kB/s]
 77% 24.0M/31.3M [00:00<00:00, 64.0MB/s]
100% 31.3M/31.3M [00:00<00:00, 89.9MB/s]
  0% 0.00/795k [00:00<?, ?B/s]
100% 795k/795k [00:00<00:00, 108MB/s]
  0% 0.00/1.87k [00:00<?, ?B/s]
100% 1.87k/1.87k [00:00<00:00, 1.60MB/s]


In [ ]:
# Unzipping
zip_file = os.listdir('dataset')
for file in zip_file:
    !unzip /content/dataset/{file}
!rm -rf /content/dataset

Archive:  /content/dataset/sample_submission.csv.zip
  inflating: sample_submission.csv   
Archive:  /content/dataset/products.csv.zip
  inflating: products.csv            
   creating: __MACOSX/
  inflating: __MACOSX/._products.csv  
Archive:  /content/dataset/order_products__train.csv.zip
  inflating: order_products__train.csv  
  inflating: __MACOSX/._order_products__train.csv  
Archive:  /content/dataset/departments.csv.zip
  inflating: departments.csv         
  inflating: __MACOSX/._departments.csv  
Archive:  /content/dataset/aisles.csv.zip
  inflating: aisles.csv              
  inflating: __MACOSX/._aisles.csv   
Archive:  /content/dataset/orders.csv.zip
  inflating: orders.csv              
  inflating: __MACOSX/._orders.csv   
Archive:  /content/dataset/order_products__prior.csv.zip
  inflating: order_products__prior.csv  
  inflating: __MACOSX/._order_products__prior.csv  


**Download intermediate prepared tables.**

In [ ]:
!kaggle datasets download -d shubhamscifi/instacart --unzip

 99% 1.46G/1.47G [00:13<00:00, 189MB/s]
100% 1.47G/1.47G [00:13<00:00, 113MB/s]


**Kaggle file uploader utility:**
- To upload intermediate tables.

In [ ]:
def kaggle_file_uploader(files,id='shubhamscifi/instacart',title='instacart',folder='data',msg='',first_time=False,del_after_upload=True):
    '''Uploads list of files to kaggle.
    Note: make sure to run after kaggle authentication.
    id : must be between 6-50 chars after "username/".
    title : must be between 6-50 chars.
    files : list of path of files that are to be uploaded.
    first_time: True if the data is being uploaded for the first time.
    del_after_upload: True if given folder needs to be deleted after file upload finishes.'''
    # https://github.com/Kaggle/kaggle-api

    # create data package json file
    !mkdir {folder}
    !kaggle datasets init -p {folder}

    # preparing metadata json file
    import json,os
    metadata = open(os.path.join(folder,'dataset-metadata.json'),'r+')
    meta = json.load(metadata)
    meta['id'] = id
    meta['title']= title
    metadata.seek(0)
    json.dump(meta,metadata)
    metadata.truncate()
    metadata.close()

    for file in set(files):
        !cp {file} {folder}

    # upload dataset to kaggle
    if (first_time):
        !kaggle datasets create -p {folder}
    else:
        # Create a New Dataset Version
        !kaggle datasets version -p {folder} -m '{msg}'

    if (del_after_upload):
        !rm -rf {folder}

**Changing the display of DataFrames as interactive tables.**

In [ ]:
# dataframes are shown as rich, interactive tables
%load_ext google.colab.data_table

In [ ]:
# restoring the standard static display
%unload_ext google.colab.data_table

## Importing libraries

In [ ]:
!pip install --upgrade plotly
!pip3 install --upgrade category_encoders
!pip install -U memory_profiler
%load_ext memory_profiler

     |████████████████████████████████| 20.6MB 68.0MB/s 
  Found existing installation: plotly 4.4.1
    Uninstalling plotly-4.4.1:
      Successfully uninstalled plotly-4.4.1
     |████████████████████████████████| 81kB 4.2MB/s 
  Created wheel for memory-profiler: filename=memory_profiler-0.58.0-cp37-none-any.whl size=30188 sha256=d3a1dca997583c514a0bc474c6da6ba1b73da6e3ed0a3f19b9af15e2b140c71f
  Stored in directory: /root/.cache/pip/wheels/02/e4/0b/aaab481fc5dd2a4ea59e78bc7231bb6aae7635ca7ee79f8ae5
Successfully built memory-profiler


In [ ]:
# requirements file
!pip freeze > requirements.txt

In [ ]:
import plotly
import plotly.express as px
import numpy as np
from scipy import stats
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
import re
import gc   # garbage collector
import pickle
# https://pypi.org/project/tqdm/#:~:text=jupyter%20console.%20Use-,auto,-instead%20of%20autonotebook
from tqdm.auto import tqdm
import time
from joblib import Parallel, delayed
from sklearn.metrics import f1_score,confusion_matrix,\
                            precision_recall_fscore_support,classification_report,\
                            accuracy_score,log_loss
from sklearn.model_selection import train_test_split

bold = lambda string: '\033[1m'+string+'\033[0m'    # for bold text
printb = lambda string: print('\033[1m'+string+'\033[0m')
# https://stackoverflow.com/questions/8924173/how-do-i-print-bold-text-in-python/8930747

## Loading Data

In [ ]:
# loading data into pandas dataframe
orders = pd.read_csv('/content/orders.csv',dtype={'order_id':np.uint32,
                                                  'user_id' :np.uint32,
                                                  'order_number':'uint8',
                                                  'order_hour_of_day':'uint8',
                                                  'order_dow':'uint8',
                                                  'days_since_prior_order':'float16'})
dep = pd.read_csv('/content/departments.csv', dtype={'department_id':'uint8',
                                                     'department': str})
aisles = pd.read_csv('/content/aisles.csv', dtype={'aisle_id':'uint8',
                                                     'aisle': str})
products = pd.read_csv('/content/products.csv', dtype={'aisle_id':'uint8',
                                                     'department_id':'uint8',
                                                     'product_name': str,
                                                     'product_id': np.uint16})
order_products_prior = pd.read_csv('/content/order_products__prior.csv',
                                   dtype={'add_to_cart_order':'uint8',
                                          'reordered':'uint8',
                                          'order_id':np.uint32,
                                          'product_id':np.uint16})
order_products_train = pd.read_csv('/content/order_products__train.csv',
                                   dtype={'add_to_cart_order':'uint8',
                                          'reordered':'uint8',
                                          'order_id':np.uint32,
                                          'product_id':np.uint16})

<a name="data-overview"></a>
## **Data overview**🔎
- Overviewing each Table
- Understanding each column.


#### **Aisles**

In [ ]:
aisles

aisle_id                       aisle
0           1       prepared soups salads
1           2           specialty cheeses
2           3         energy granola bars
3           4               instant foods
4           5  marinades meat preparation
..        ...                         ...
129       130    hot cereal pancake mixes
130       131                   dry pasta
131       132                      beauty
132       133  muscles joints pain relief
133       134  specialty wines champagnes

[134 rows x 2 columns]

In [ ]:
aisles.loc[aisles.aisle.str.contains('missing'),:]

aisle_id    aisle
99       100  missing

In [ ]:
aisles.info(verbose=True, memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   aisle_id  134 non-null    uint8 
 1   aisle     134 non-null    object
dtypes: object(1), uint8(1)
memory usage: 9.7 KB


> **Observations:**
- Total 134 aisles.
- aisle number 100 is not missing but it's name is "missing".
- Maybe, it is the aisle for missing or non-assigned products.

---



#### **Departments**

In [ ]:
dep

department_id       department
0               1           frozen
1               2            other
2               3           bakery
3               4          produce
4               5          alcohol
5               6    international
6               7        beverages
7               8             pets
8               9  dry goods pasta
9              10             bulk
10             11    personal care
11             12     meat seafood
12             13           pantry
13             14        breakfast
14             15     canned goods
15             16       dairy eggs
16             17        household
17             18           babies
18             19           snacks
19             20             deli
20             21          missing

In [ ]:
dep.loc[dep.department.str.contains('missing'),:]

department_id department
20             21    missing

In [ ]:
dep.info(verbose=True, memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   department_id  21 non-null     uint8 
 1   department     21 non-null     object
dtypes: object(1), uint8(1)
memory usage: 1.5 KB


> **Observations:**
- Total 21 departments.
- Similar, to aisle number 100 department number 21's name is also "missing".
- Maybe, it is the department for missing or non-assigned products.

---

#### **Products**

In [ ]:
products

product_id  ... department_id
0               1  ...            19
1               2  ...            13
2               3  ...             7
3               4  ...             1
4               5  ...            13
...           ...  ...           ...
49683       49684  ...             5
49684       49685  ...             1
49685       49686  ...             3
49686       49687  ...             8
49687       49688  ...            11

[49688 rows x 4 columns]

In [ ]:
products.info(verbose=True,memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49688 entries, 0 to 49687
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     49688 non-null  int64 
 1   product_name   49688 non-null  object
 2   aisle_id       49688 non-null  uint8 
 3   department_id  49688 non-null  uint8 
dtypes: int64(1), object(1), uint8(2)
memory usage: 4.6 MB


>**Observations:**
- Total 49,688 (~50k) products.
- Each product belongs to a unique aisle and department.

---



#### **Orders**

In [ ]:
orders

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0         2539329        1  ...                 8                     NaN
1         2398795        1  ...                 7                    15.0
2          473747        1  ...                12                    21.0
3         2254736        1  ...                 7                    29.0
4          431534        1  ...                15                    28.0
...           ...      ...  ...               ...                     ...
3421078   2266710   206209  ...                18                    29.0
3421079   1854736   206209  ...                10                    30.0
3421080    626363   206209  ...                12                    18.0
3421081   2977660   206209  ...                12                     7.0
3421082    272231   206209  ...                14                    30.0

[3421083 rows x 7 columns]

In [ ]:
orders.info(null_counts=True, memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   order_id                3421083 non-null  int64  
 1   user_id                 3421083 non-null  int64  
 2   eval_set                3421083 non-null  object 
 3   order_number            3421083 non-null  uint8  
 4   order_dow               3421083 non-null  uint8  
 5   order_hour_of_day       3421083 non-null  uint8  
 6   days_since_prior_order  3214874 non-null  float16
dtypes: float16(1), int64(2), object(1), uint8(3)
memory usage: 270.7 MB


In [ ]:
print('No. of unique values per column:')
print('--------------------------------')
orders.nunique()

No. of unique values per column:
--------------------------------


order_id                  3421083
user_id                    206209
eval_set                        3
order_number                  100
order_dow                       7
order_hour_of_day              24
days_since_prior_order         31
dtype: int64

> **Observation:**
- Total 3,421,083 (~3.4M) orders.
- Total 206,209 (~206k) users.
- <u>order_number</u>: the order sequence number for this user (1 = first, n = nth).
- <u>order_dow</u>: the day of the week the order was placed on (0-6).
- <u>order_hour_of_day</u>: the hour of the day the order was placed on (0-23).

In [ ]:
orders.eval_set.unique()

array(['prior', 'train', 'test'], dtype=object)

In [ ]:
print('Each order belongs to either train or test or prior set.')
print('\t\t user_id=1')
display(orders.loc[orders.user_id==1,['order_id','user_id','eval_set','order_number']].head(20))
print('\n\t\t user_id=32')
display(orders.loc[orders.user_id==32,['order_id','user_id','eval_set','order_number']].head(20))
print('Each user belongs to either train or test set.')
print("""Depending upon the eval_set value of the user's last order\
 i.e. train/test the user belongs to that set respectively.""")

Each order belongs to either train or test or prior set.
		 user_id=1


order_id  user_id eval_set  order_number
0    2539329        1    prior             1
1    2398795        1    prior             2
2     473747        1    prior             3
3    2254736        1    prior             4
4     431534        1    prior             5
5    3367565        1    prior             6
6     550135        1    prior             7
7    3108588        1    prior             8
8    2295261        1    prior             9
9    2550362        1    prior            10
10   1187899        1    train            11


		 user_id=32


order_id  user_id eval_set  order_number
470    134879       32    prior             1
471   1275838       32    prior             2
472   1709286       32    prior             3
473     72626       32    prior             4
474   1210985       32    prior             5
475   3202221       32     test             6

Each user belongs to either train or test set.
Depending upon the eval_set value of the user's last order i.e. train/test the user belongs to that set respectively.


In [ ]:
print('Number of orders belonging to each set:')
orders.value_counts('eval_set')

Number of orders belonging to each set:


eval_set
prior    3214874
train     131209
test       75000
dtype: int64

> **Observaitons:**
- <u>eval_set</u>: which evaluation set this order belongs in (train/test/prior).
    - "prior": orders prior to that user's most recent order (~3.2m orders)
    - "train": user's most recent order for training (~131k orders)
    - "test": user's most recent order test data for which we have to do prediction (75k orders)
    - ![](https://i.imgur.com/S0Miw3m.png) 
    - img src:
    https://www.kaggle.com/kokovidis/ml-instacart-f1-0-38-part-two-xgboost-f1-max
- Each user is associated with either train or test data, since his/her last order is in either train or test eval_set.
    - ~131k train users.
    - 75k test users.

In [ ]:
print('No. of null values in days_since_prior_order column = ', end='')
orders.days_since_prior_order.isna().sum()

No. of null values in days_since_prior_order column = 

206209

> **Observations:**
- <u>days_since_prior_order</u>: days since the last order, capped at 30 (with NaNs for order_number = 1)
    - days_since_prior_order col. contains 206,209 (~206k) NaN values.
    - These are not missing values.
    - There is no prior order for the 1st order placed by an user. Hence, this column's value for such cases are kept NaN. That's why no. of NaN values is exactly equal to no. of users.
- Basically, this table contains the metadata of each order.

---



#### **order_products_prior**

In [ ]:
order_products_prior

order_id  product_id  add_to_cart_order reordered
0                2       33120                  1         1
1                2       28985                  2         1
2                2        9327                  3         0
3                2       45918                  4         1
4                2       30035                  5         0
...            ...         ...                ...       ...
32434484   3421083       39678                  6         1
32434485   3421083       11352                  7         0
32434486   3421083        4600                  8         0
32434487   3421083       24852                  9         1
32434488   3421083        5020                 10         1

[32434489 rows x 4 columns]

In [ ]:
order_products_prior.info(memory_usage='deep', null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Non-Null Count     Dtype   
---  ------             --------------     -----   
 0   order_id           32434489 non-null  int64   
 1   product_id         32434489 non-null  int64   
 2   add_to_cart_order  32434489 non-null  uint8   
 3   reordered          32434489 non-null  category
dtypes: category(1), int64(2), uint8(1)
memory usage: 556.8 MB


> **Observations:**
- This table contains all the products purchased in each prior order with the sequence of products purchased in that order. For both train and test users.
- <u>add_to_cart_order</u>: order in which each product was added to cart.
- <u>reordered</u>: 1 if this product has been ordered by the user in the past, 0 otherwise.
    - 1 - 'reorder'
    - 0 - 'first purchase'

---



#### **order_products_train**

In [ ]:
order_products_train

order_id  product_id  add_to_cart_order  reordered
0               1       49302                  1          1
1               1       11109                  2          1
2               1       10246                  3          0
3               1       49683                  4          0
4               1       43633                  5          1
...           ...         ...                ...        ...
1384612   3421063       14233                  3          1
1384613   3421063       35548                  4          1
1384614   3421070       35951                  1          1
1384615   3421070       16953                  2          1
1384616   3421070        4724                  3          1

[1384617 rows x 4 columns]

In [ ]:
order_products_train.info(memory_usage='deep', null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1384617 entries, 0 to 1384616
Data columns (total 4 columns):
 #   Column             Non-Null Count    Dtype
---  ------             --------------    -----
 0   order_id           1384617 non-null  int64
 1   product_id         1384617 non-null  int64
 2   add_to_cart_order  1384617 non-null  uint8
 3   reordered          1384617 non-null  int8 
dtypes: int64(2), int8(1), uint8(1)
memory usage: 23.8 MB


> **Observations:**
- This table contains all the products purchased in each train order with the sequence of products purchased in that order.
- <u>add_to_cart_order</u>: order in which each product was added to cart.
- <u>reordered</u>: 1 if this product has been ordered by the user in the past, 0 otherwise.
    - 1 - 'reorder'
    - 0 - 'first purchase'
- There are no missing values in any of the tables.

---



## **Mapping to a Supervised classification problem:**
- Task: Given order_id i.e.{user info + info of prior products purchased by that user} predict all the products that will be reordered.
- Since, we have to predict multiple products for a given order_id. This is a multilabel classification task with (~50k) labels.
- Simplifying the task:
    - Given the order_id, we know the user_id.
    - The products that the user has never bought can't be reordered.
    - Only the prior products(i.e. products bought earlier) can be reordered by this user.
    - We can run **binary classification** for all such products, where "reordered" is target variable(Y).
    - Predictor variables(X) will contain user_info & product_info.
    - Hence the evaluation metric(mean F1-score) can be calculated using the F1-score on binary classification task.

![](https://i.imgur.com/Yb1CKAF.png)

img src: https://www.kaggle.com/kokovidis/ml-instacart-f1-0-38-part-two-xgboost-f1-max

---



## Data Cleaning.🧹
- Deduplication
- Find and Fix any missing values.
- Find and Fix any Garbage/incorrect values.

#### **Deduplication**

In [ ]:
print('Is there any duplicate rows in aisles table?','->',sep='\n',end='')
aisles.duplicated().any()

Is there any duplicate rows in aisles table?
->

False

In [ ]:
print('Is there any duplicate rows in departments table?','->',sep='\n',end='')
dep.duplicated().any()

Is there any duplicate rows in departments table?
->

False

In [ ]:
print('Is there any duplicate rows in products table?','->',sep='\n',end='')
products.duplicated().any()

Is there any duplicate rows in products table?
->

False

In [ ]:
print('Is there any duplicate rows in orders table?','->',sep='\n',end='')
orders.duplicated().any()

Is there any duplicate rows in orders table?
->

False

In [ ]:
print('Is there any duplicate rows in order_products_prior table?','->',sep='\n',end='')
order_products_prior.duplicated().any()

Is there any duplicate rows in order_products_prior table?
->

False

In [ ]:
print('Is there any duplicate rows in order_products_train table?','->',sep='\n',end='')
order_products_train.duplicated().any()

Is there any duplicate rows in order_products_train table?
->

False

> **Observations:**
- There are no duplicates in any of the tables.

---



#### **Missing Values**
- As we saw in the [Data overview](#data-overview) section, there are no missing values in any of the tables.

---



#### **Garbage Values**

In [ ]:
print('       aisles table')
print('  Column         Is there any garbage value?')
print('-----------      ---------------------------')
sep='          ' 
print('aisle_id',aisles.aisle_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('   aisle',aisles.aisle.str.contains('[^a-zA-Z ]').any(),sep=sep)

       aisles table
  Column         Is there any garbage value?
-----------      ---------------------------
aisle_id          False
   aisle          False


In [ ]:
print('       departments table')
print('    Column         Is there any garbage value?')
print('-------------      ---------------------------')
sep='          ' 
print('department_id',dep.department_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('   department',dep.department.str.contains('[^a-zA-Z ]').any(),sep=sep)

       departments table
    Column         Is there any garbage value?
-------------      ---------------------------
department_id          False
   department          False


In [ ]:
print('         products table')
print('    Column         Is there any garbage value?')
print('-------------      ---------------------------')
sep='          ' 
print('   product_id',products.product_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('     aisle_id',products.product_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('department_id',products.product_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print(' product_name',products.product_name.str.contains('[^a-zA-Z ]').any(),sep=sep)

         products table
    Column         Is there any garbage value?
-------------      ---------------------------
   product_id          False
     aisle_id          False
department_id          False
 product_name          True


In [ ]:
printb('Special characters in product names:')
special_char = set(re.findall("[^a-zA-Z0-9 ]",''.join(products.loc[:,'product_name'])))
print(special_char)
print(f'Total {len(special_char)} distinct special characters.')

Special characters in product names:
{'#', '/', 'ó', '™', 'í', '’', '´', '\x8f', ')', '"', 'û', '?', 'á', '�', '*', 'â', 'ú', '&', '\\', '.', '%', 'е', '(', '}', '˚', '°', '-', '‑', 'é', ',', '”', 'è', 'ñ', '–', 'ü', '`', '{', '®', '=', 'ô', '+', '!', ':', '@', '$', '\xa0', "'", 'ç', ';'}
Total 49 distinct special characters.


In [ ]:
print("\033[1m","Let's see some product names containing above special chars:","\033[0m") # \033[1m for printing bold
print("------------------------------------------------------------")
print('char                      product name')
print('----         ----------------------------------------')
sep= '          '
for char in special_char:
    print(repr(char),products.loc[products.product_name.str.contains(char,regex=False),'product_name'].iloc[0], sep=sep)

 Let's see some product names containing above special chars: 
------------------------------------------------------------
char                      product name
----         ----------------------------------------
'#'          Natural Brown Cone Style Coffee Filters #4
'/'          24/7 Performance Cat Litter
'ó'          Chile Limón Flavored Potato Chips
'™'          Mouthwash Clean Mint Listerine® Zero™
'í'          Blueberry Pomegranate Acaí Cultured Goat Milk Kefir
'’'          Kickin’ BBQ Chopped Salad
'´'          Women´s Multivitamin Gummies
'\x8f'          Hershey's Chocolate Creme Pie Singles
')'          Fabric Refresher Meadows & Rain Air Freshener (1 Count, 27 oz)  Air Care
'"'          Easy Grab 9\"x13\" Oblong Glass Bakeware
'û'          Milk Chocolate Truffle Brûlée
'?'          Ksa Ko?Lsch Style Ale
'á'          Reserve Dark Roast Guatemala San Sebastián Ground Coffee
'�'          Diana�s Bananas Milk Chocolate with Peanuts - Banana Babies � 5 CT
'*'          Fresh 

In [ ]:
# Remove special characters from product names.
x=products.product_name.astype(str)\
                    .replace({'\xa0': '',
                               '\x8f': '', 
                               r'[{}]': '',
                               r'\*': '',
                               r'e': 'ea',
                               '�': '',
                               '[-–‑]': ' ',
                               r';': '',
                               '=': ' equal to ',
                               '@': 'at',
                               '\+': ' plus ',
                               '[âá]': 'a',
                               '[еéè]': 'e',
                               'í': 'i',
                               'ñ': 'n',
                               '[óô]': 'o',
                               '[üûú]': 'u',
                               'ç': 'c',
                               '[°˚]': ' degree ',
                               '[®™]': '',
                               r"'s": '',
                               r'`s': '',
                               r'´s': '',
                               r"'": '',
                               r"’": '',
                               r"”": ' inch ',
                               r'\"': ' inch ',
                               r'\\': '',
                               ':': '',
                               '\?': ' ',
                               r',': '',
                               r'%': ' percent ',
                               r'&#':' No. ',
                               r'#': ' No. ',
                               r' & ': ' and ',
                               r'&': ' and ',
                               r'!': '',
                               r'24/7': ' 24_7 ',
                               r'w/': ' with ',
                               r'/': ' ',
                               r'\$': ' Dollar ',
                               'Bananas': 'Banana',
                               'Strawberries': 'Strawberry',
                               r'( )+': ' '},
                              regex=True)
x = x.str.strip()   # Removing trailing and leading whitespaces

In [ ]:
products['product_name'] = x
del x
products

product_id  ... department_id
0               1  ...            19
1               2  ...            13
2               3  ...             7
3               4  ...             1
4               5  ...            13
...           ...  ...           ...
49683       49684  ...             5
49684       49685  ...             1
49685       49686  ...             3
49686       49687  ...             8
49687       49688  ...            11

[49688 rows x 4 columns]

"^a-zA-Z0-9:@$&,%!?.âáеéèíñóôüûúç°˚®™´'’”=+/()#\"\-–‑ �\\\;{}`*\xa0\x8f"
- All unnecessary characters have been removed.

In [ ]:
print('                orders table')
print('         Column              Is there any garbage value?')
print('-----------------------      ---------------------------')
sep='          ' 
print('               order_id',orders.order_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('                user_id',orders.user_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('               eval_set',orders.eval_set.astype('str').str.contains('[^\b(prior|train|test)\b]').any(),sep=sep)
print('           order_number',orders.order_number.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('              order_dow',not(orders.order_dow.between(0,6).any()),sep=sep)
print('      order_hour_of_day',not(orders.order_hour_of_day.between(0,23).any()),sep=sep)
print(' days_since_prior_order',not(orders.days_since_prior_order.between(0,30).any()),sep=sep)

                orders table
         Column              Is there any garbage value?
-----------------------      ---------------------------
               order_id          False
                user_id          False
               eval_set          False
           order_number          False
              order_dow          False
      order_hour_of_day          False
 days_since_prior_order          False


In [ ]:
print('       order_products_prior table')
print('      Column          Is there any garbage value?')
print('-----------------     ---------------------------')
sep='          ' 
print('         order_id',order_products_prior.order_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('       product_id',order_products_prior.product_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('add_to_cart_order',order_products_prior.add_to_cart_order.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('        reordered',not(order_products_prior.reordered.astype('uint8').between(0,1).any()),sep=sep)

       order_products_prior table
      Column          Is there any garbage value?
-----------------     ---------------------------
         order_id          False
       product_id          False
add_to_cart_order          False
        reordered          False


In [ ]:
print('       order_products_train table')
print('      Column           Is there any garbage value?')
print('-----------------      ---------------------------')
sep='          ' 
print('         order_id',order_products_train.order_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('       product_id',order_products_train.product_id.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('add_to_cart_order',order_products_train.add_to_cart_order.astype('str').str.contains('[^0-9]').any(),sep=sep)
print('        reordered',not(order_products_train.reordered.astype('uint8').between(0,1).any()),sep=sep)

       order_products_train table
      Column           Is there any garbage value?
-----------------      ---------------------------
         order_id          False
       product_id          False
add_to_cart_order          False
        reordered          False




---

